File: Third_semestr/Homework_2/ModelOfLAN/ModelOfLAN/LocalNetwork.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ModelOfLAN
{
    /// <summary>
    /// this class simulates the work of the local network
    /// </summary>
    public class LocalNetwork
    {
        /// <summary>
        /// class constructor
        /// </summary>
        /// <param name="file">the adjacency matrix</param>
        /// <param name="infectedComputers">computers that are infected with viruses</param>
        public LocalNetwork(System.IO.StreamReader file, List<int> infectedComputers)
        {
            computers = new List<Computer>();
            string line;
            if ((line = file.ReadLine()) != null)
            {
                for (int j = 0; j < line.Length; ++j)
                {
                    computers.Add(new Computer(line[j]) { ID = j });
                }
            }

            adjacencyList = new List<List<int>>();
            int i = 0;
            while ((line = file.ReadLine()) != null)
            {
                adjacencyList.Add(new List<int>());
                for (int j = 0; j < line.Length; ++j)
                {
                    if (line[j] == '1')
                    {
                        adjacencyList[i].Add(j);
                    }
                }

                ++i;
            }

            foreach (var numOfInfected in infectedComputers)
            {
                computers[numOfInfected].Infect(100);
            }
        }

        /// <summary>
        /// shows the adjacency matrix in the console
        /// </summary>
        public void ShowAdjacencyList()
        {
            foreach (var list in adjacencyList)
            {
                foreach (var computer in list)
                {
                    Console.Write(computer + " ");
                }

                Console.WriteLine();
            }
        }

        /// <summary>
        /// checks the health of the system
        /// </summary>
        /// <returns>returns true, if system is unhealthy</returns>
        public bool IsUnhealthy() => computers.Aggregate(true, (current, computer) => current && computer.IsInfected());

        /// <summary>
        /// shows the health of each computer
        /// </summary>
        public void ShowSystem()
        {
            foreach (var computer in computers)
            {
                Console.ForegroundColor = computer.IsInfected() ? ConsoleColor.Red : ConsoleColor.Gray;
                var health = computer.IsInfected() ? "infected" : "healthy";
                Console.WriteLine("computer number " + computer.ID + " is " + health);   
            }
        }

        /// <summary>
        /// simulates the unit of time and performs changes in the system
        /// </summary>
        public void Step()
        {
            var tryInfect = new List<int>();
            for (int i = 0; i < adjacencyList.Count; ++i)
            {
                if (computers[i].IsInfected())
                {
                    foreach (var neighbors in adjacencyList[i])
                    {
                        tryInfect.AddRange(adjacencyList[i]);
                    }
                }
            }

            var randomDamage = new Random().Next(100);
            foreach (var num in tryInfect)
            {
                computers[num].Infect(randomDamage);
            }
        }

        /// <summary>
        /// returns a list of healthy computers
        /// </summary>
        /// <param name="list">list for copy</param>
        /// <returns></returns>
        public List<int> GetHealthyComputers()
        {
            var list = new List<int>();
            foreach (var computer in computers)
            {
                if (!computer.IsInfected())
                {
                    list.Add(computer.ID);
                }
            }

            return list;
        }

        private List<List<int>> adjacencyList;
        private List<Computer> computers;
    }
}

File: Fourth_semestr/Homework_6/ModelOfLAN/ModelOfLAN/LocalNetwork.fs
namespace ModelOfLAN

open System.IO

/// type of operating system
type OS = | Windows | Linux | OS_X

/// computer of local network
type Computer(abbr:char, id:int) =
    //private
    let mutable isInfected = false
    let healthPoints =
        match abbr with
        | 'w' -> 60
        | 'l' -> 70
        | 'o' -> 50
        | _ -> 30

//    /// type of operating system
//    member val OperatingSystem = os with get, set
    
    /// ID of computer in local network
    member val ID = id with get
    
    /// checks the health of computer
    member this.IsInfected () = isInfected;
    
    /// simulates a viral attack on the computer
    member this.Infect damagePoint =
        if damagePoint >= healthPoints then
            isInfected <- true


/// this class simulates the work of the local network
type LocalNetwork(file:string, infectedComputers) =
    let mutable computers = Array.empty
    let mutable adjacencyList = Seq.empty
    
    do
        use stream = new StreamReader(file)
        computers <- Seq.toArray (Seq.mapi (fun i x -> new Computer(x, i)) <| stream.ReadLine())
        Seq.iter (fun i -> computers.[i].Infect(100)) infectedComputers

        adjacencyList <-
            seq { use stream = new StreamReader(file)
                  ignore <| stream.ReadLine()
                  while not stream.EndOfStream do
                    let line = stream.ReadLine()
                    let neighbors = Seq.map (fun x -> int x)                //Seq.fold (fun acc x -> (int x) :: acc) []
                                    << String.filter (fun x -> x <> '0')             ////???????????????????
                                    <| String.mapi (fun i x -> if (x = '1') then char i else x) line
                    yield neighbors }
    
    /// shows the adjacency matrix in the console
    member this.ShowAdjList () =
        Seq.iter (fun ls -> Seq.iter (fun n -> printf "%A" n
                                               printf " ") ls
                            printfn "") adjacencyList


    /// shows the health of each computer
    member this.ShowSystem () =
        let fontColor (comp:Computer) =
            if comp.IsInfected () then System.Console.ForegroundColor <- System.ConsoleColor.Red
            else System.Console.ForegroundColor <- System.ConsoleColor.Gray
        let health (comp:Computer) =
            if comp.IsInfected () then "infected"
            else "healthy"
        Seq.iter (fun x -> fontColor x
                           System.Console.WriteLine("computer number " + x.ID.ToString () + " is " + (health x))) computers
    

    /// simulates the unit of time and performs changes in the system
    member this.Step () =
        let infected =
            Array.fold
                (fun acc (comp:Computer) -> if comp.IsInfected () then comp.ID :: acc else acc)
                []
                computers
        let tryInfect =
            Seq.fold (fun acc i -> Seq.append acc <| Seq.item i adjacencyList) Seq.empty infected
        let genRandomNumbers =
            let rnd = System.Random()
            List.init computers.Length (fun _ -> rnd.Next ())
        Seq.iter2 (fun i dmg -> computers.[i].Infect(dmg)) tryInfect genRandomNumbers

    /// checks the health of the system
    member this.IsUnhealthy () =
        Array.forall (fun (comp:Computer) -> comp.IsInfected ()) computers

    /// returns a list of healthy computers
    member this.GetHealthyComputers () =
        Array.fold (fun acc (comp:Computer) -> if comp.IsInfected () then (comp.ID :: acc) else acc) [] computers